Update CONTRIBUTING.md to reflect current reality (#1354)
nginx/unit: CONTRIBUTING.md

To suggest an enhancement, open an
[issue](https://github.com/nginx/unit/issues/new) on GitHub with the label
`enhancement`.  Please do this before implementing a new feature to discuss
the feature first.

### Open a Pull Request

Before submitting a PR, please read the NGINX Unit code guidelines to know
more about coding conventions and benchmarks.  Fork the repo, create a branch,
and submit a PR when your changes are tested and ready for review.  Again, if
you'd like to implement a new feature, please consider creating a feature
request issue first to start a discussion about the feature.

## Git Style Guide

> - Create atomic commits.  A commit should do just one thing, I.e you

Reviewer: Double space, and i.e. is lowercase when in the middle of the sentence, like here, and because it's an abbreviation, the periods are needed after both i and e. See https://www.grammarly.com/blog/know-your-latin-i-e-vs-e-g/
```suggestion
- Create atomic commits. A commit should do just one thing, i.e. you
```

Reviewer: Is there a reason for that practice in this file? And I assume other files. I understand that double spaces were important when typewriters were around and I'm missing context on whether that or some similar reason made its way to here.

I also want to gently push back on any explanation that sounds like "we've always done it this way." In those cases I'd like to either figure out what the original reason was, see if that's still relevant, and potentially change it if it no longer serves us. 🙂

Author: I've no idea, probably personal writing style.

>   shouldn't mix whitespace changes with actual code changes.  Do the

Reviewer: I woulc argue that whitespace changes are also actual code changes. This makes it feel like whitespace changes are somehow lesser than "real" code changes, which isn't something we should allude to. Maybe functional changes?
```suggestion
  shouldn't mix whitespace changes with functional code changes.  Do the
```

Author: It of course depends on the language...

In C, whitespace is not significant, whereas in Python it is... seeing as we're a mostly C project, I'd say the above holds, but yes we do have some other language code where whitespace _matters_ e.g Python, so I'll change as suggested.

>   whitespace changes in a separate commit first.

Reviewer: I don't think the order is important.
```suggestion
  whitespace changes in a separate commit.
```

Author: I disagree.

In general you should do any cleanup patches first, e.g as part of preparatory patches for a change.

It's the same as fixing a bunch of coding style issues on some code _before_ you make actual changes to _that_ code.

Reviewer: I still think cleanup can come as a later step as long as that's also included. I don't think being prescriptive about having to do cleanup before anything happens is useful. As long as cleanup happens, why should we care in what order that happened?

Sure, individual preferences exist, but those are individual, and we shouldn't push those on others.

Author: Except for when you're the one doing the review, I know I'm not the only one that would ask for it to be done that way...

The ordering is somewhat logical. You do preparatory work before _actual_ work.

Let's assume the whitespace changes are trivial and uncontentious, if you _did_ them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

It may also help with things like git-blame(1).

Author: `whitespace` changes was just an example, replace that with `refactor`, then tell me that shouldn't come first...

Reviewer: It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.

Others might want to add the new / changed functionality first, and then tidy it up.

Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

I think we should be okay with all kinds of workflows and not prescribe a rigid order.

Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

Reviewer: > if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.

I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

Author: > It depends. I would treat refactor specifically as its own PR. My argument here is that there are people who want to clean up the whitespace issues / refactor needs first, and then add the new / changed functionality.
> 
> Others might want to add the new / changed functionality first, and then tidy it up.
> 
> Yet others might want to do cleanup first, add new / changed functionality after, and then tidy it up even more.

If you do things right there shouldn't be any _need_ to _tidy_ things up afterwards. I.e. this is bad form

```
Fixup previous commit
Introduce new feature
```

All it does is add extra noise.

If you do your refactor/cleanup first then it is often the case that your subsequent commits are _simpler_ and _more concise_.

You want commits to be as small and concise as possible. This really helps with reviewing for one thing.

> I think we should be okay with all kinds of workflows and not prescribe a rigid order.
> 
> Everyone should be free to do it in whatever order they see fit as long as the end result passes lint checks, and the individual commits are atomic.

Not really.

At the end of day it's up to the project maintainers how things are done. That is all this commit is doing, updating the contributing document to reflect how things are _currently_ done in Unit.

If you disagree with how the Unit project is currently run then that is a separate conversation.

> > if you did them after the real work and you need to change the real work you may have to also redo the whitespace changes, at the very least these commits will get new id's and just add general noise.
> 
> I agree, at the same time we're also asking folks to rebase on their locals and force push anyways, so every time there's a change, most of the commits are going to be totally new anyways. If they also mark the whitespace

Only from the point of the rebase...

> change commits with a `whitespace` prefix, then as a reviewer we know we can skip that one.

In Unit (like in many other projects) the `prefix` has a specific purpose and saying what _kind_ of commit it is isn't it... also I hope you didn't just say, "_no need to review whitespace changes_"... because they should be reviewed along with everything else.

[ Also just because a patch might be labelled as `whitespace` doesn't mean there won't be _other_ changes in there! ]

  Ideally you should rebase locally and force push new commits up.

> - In the subject line, use the imperative mood.  I.e write the subject like

Reviewer: Double space, missing period after e. I.e. is at the beginning of the sentence, so capital I is good here.
```suggestion
- In the subject line, use the imperative mood. I.e. write the subject like
```

Author: Again following pre-existing style... heh, I've never wrote these things with a trailing period...

>   you're giving git a command, e.g "Free memory before exiting". Do _not_

Reviewer: E.g. is also an abbreviation, thus both periods are needed. Punctuation mark in this case goes inside the quotation marks, see https://www.grammarly.com/blog/quotation-marks/

Highlighting the _not_ is a bit harsh in my opinion, we can be friendlier than that. 🙂 
```suggestion
  you're giving git a command, e.g. "Free memory before exiting." Do not
```

Author: This is an example of a commit subject and the '.' is _not_ part of the commit subject.

Reviewer: Good point, I retract this changeset, thanks!

  terminate the subject with a `.`

- Try to limit the subject line to around 50 characters, but try not to
  exceed 72.

- Wrap the body of the commit message after 72 characters.

- Use lowercase subject line prefixes for commits that affect a specific
  portion of the code; examples include "tests:", "ci:", or "http:", and

>   also individual languages such as "python:" or "php:".  Also if multiple

Reviewer: Repetition: also, rewrite, continuing on new line
```suggestion
  also individual languages such as "python:" or "php:". Combine
```

>   areas are affected, e.g "auto, perl:"

Reviewer: ```suggestion
  prefixes if multiple areas are affected, e.g. "auto, perl:"
```

Author: Hmm, that suggestion doesn't make sense to me...

Author: I've _tweaked_ it anyway...

Reviewer: This suggestion goes hand in hand with the Combine one above this. Individually they won't make sense.

- If the commit fixes an open issue then you can use the "Closes:"
  tag/trailer to reference it and have GitHub automatically close it once

>   it's been merged.  E.g

Reviewer: ```suggestion
  it's been merged.  E.g.:
```

  `Closes: https://github.com/nginx/unit/issues/9999`

That should go at the end of the commit message, separated by a blank line,

> along with any other tags.

Reviewer: Not a blocker. I understand this belongs with the previous point about the closes tag. Does this need to be indented to render as being part of the bullet point?

Author: Hmm, I think it's OK the way it is, it's indented at the same level of the bullet point it's part of and if there were more bullet points after it I think it'd still be clear what it belongs to...

Reviewer: I might be missing something. When I go to view this file at this commit (https://github.com/nginx/unit/blob/db59b52964e209fb1d954a1a93bb1481ccca3f8e/CONTRIBUTING.md), this is what it renders as:

![image](https://github.com/user-attachments/assets/1a05c032-7820-4c53-bf7c-9b85e590b128)

It looks like that lost line isn't part of the last bullet point.

Author: Yes, that's what I see...

Author: Oh, the _last_ line, yes that should perhaps be indented...
